kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/ProxyChannelStateMachine.java

```diff
@@ -0,0 +1,447 @@
+ *      ↓ frontend.{@link KafkaProxyFrontendHandler#channelRead(ChannelHandlerContext, Object) channelRead} receives any other KRPC request
+ *     {@link ProxyChannelState.SelectingServer SelectingServer} ╌╌╌╌⤍ <b>error</b> ╌╌╌╌⤍
+ *      │
+ *      ↓ netFiler.{@link NetFilter#selectServer(NetFilter.NetFilterContext) selectServer} calls frontend.{@link KafkaProxyFrontendHandler#initiateConnect(HostPort, List) initiateConnect}
+ *     {@link ProxyChannelState.Connecting Connecting} ╌╌╌╌⤍ <b>error</b> ╌╌╌╌⤍
+ *      │
+ *      ↓
+ *     {@link Forwarding Forwarding} ╌╌╌╌⤍ <b>error</b> ╌╌╌╌⤍
+ *      │ backend.{@link KafkaProxyBackendHandler#channelInactive(ChannelHandlerContext) channelInactive}
+ *      │ or frontend.{@link KafkaProxyFrontendHandler#channelInactive(ChannelHandlerContext) channelInactive}
+ *      ↓
+ *     {@link Closing Closing} ⇠╌╌╌╌ <b>error</b> ⇠╌╌╌╌⤍
+ *      │
+ *      ↓
+ *     {@link Closed Closed} ⇠╌╌╌╌ <b>error</b> ⇠╌╌╌╌
+ * </pre>
+ *
+ * <p>In addition to the "session state" this class also manages a second state machine for
+ * handling TCP backpressure via the {@link #clientReadsBlocked} and {@link #serverReadsBlocked} field:</p>
+ *
+ * <pre>
+ *     bothBlocked ←────────────────→ serverBlocked
+ *         ↑                                ↑
+ *         │                                │
+ *         ↓                                ↓
+ *    clientBlocked ←───────────────→ neitherBlocked
+ * </pre>
+ * <p>Note that this backpressure state machine is not related to the
+ * session state machine: in general backpressure could happen in
+ * several of the session states.</p>
+ */
+public class ProxyChannelStateMachine {
+    private static final Logger LOGGER = getLogger(ProxyChannelStateMachine.class);
+
+    /**
+     * The current state. This can be changed via a call to one of the {@code on*()} methods.
+     */
+    @Nullable
+    private ProxyChannelState state;
+
+    /*
+     * The netty autoread flag is volatile =>
+     * expensive to set in every call to channelRead.
+     * So we track autoread states via these non-volatile fields,
+     * allowing us to only touch the volatile when it needs to be changed
+     */
+    @VisibleForTesting
+    boolean serverReadsBlocked;
+    @VisibleForTesting
+    boolean clientReadsBlocked;
+
+    /**
+     * The frontend handler. Non-null if we got as far as ClientActive.
+     */
+    @NonNull
+    private KafkaProxyFrontendHandler frontendHandler;
+
+    /**
+     * The backend handler. Non-null if {@link #onNetFilterInitiateConnect(HostPort, List, VirtualCluster, NetFilter)}
+     * has been called
+     */
+    @VisibleForTesting
+    @Nullable
+    private KafkaProxyBackendHandler backendHandler;
+
+    @VisibleForTesting
+    ProxyChannelState state() {
+        return state;
+    }
+
+    @VisibleForTesting
+    void forceState(@NonNull ProxyChannelState state, @NonNull KafkaProxyFrontendHandler frontendHandler, @Nullable KafkaProxyBackendHandler backendHandler) {
```

Comment:
This is a test only method and not intended for production use, right?  I think it'd be good to javadoc that.

Reply:
Yes, its very much test only code. Will clarify.

Reply:
resolved in 142df9b997ed036b3cfb81d8a78003312deb99c7